src/Eshopworld.Messaging/Messenger.cs

```diff
@@ -0,0 +1,189 @@
+    using System.Collections.Generic;
+    using System.Reactive;
+    using System.Reactive.Linq;
+    using System.Reactive.Subjects;
+    using System.Threading.Tasks;
+    using Core;
+    using JetBrains.Annotations;
+
+    /// <summary>
+    /// The main messenger entry point.
+    /// It is recommended that only one messenger exists per application so you should always give this a singleton lifecycle.
+    /// </summary>
+    /// <remarks>
+    /// Some checks are only valuable if you have a unique messenger class in your application. If you set it up as transient
+    /// then some checks will only run against a specific instance and won't really validate the entire behaviour.
+    /// </remarks>
+    public class Messenger : IMessenger, IReactiveMessenger
+    {
+        internal readonly object Gate = new object();
+        internal readonly string ConnectionString;
+        internal readonly string SubscriptionId;
+
+        internal readonly ISubject<object> MessagesIn = new Subject<object>();
+
+        internal Dictionary<Type, IDisposable> MessageSubs = new Dictionary<Type, IDisposable>();
+        internal Dictionary<Type, MessageQueueAdapter> QueueAdapters = new Dictionary<Type, MessageQueueAdapter>();
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="Messenger"/>.
+        /// </summary>
+        /// <param name="connectionString">The Azure Service Bus connection string.</param>
+        /// <param name="subscriptionId">The subscription ID where the service bus namespace lives.</param>
+        public Messenger([NotNull]string connectionString, [NotNull]string subscriptionId)
+        {
+            ConnectionString = connectionString;
+            SubscriptionId = subscriptionId;
+        }
+
+        /// <inheritdoc />
+        public async Task Send<T>(T message)
+            where T : class
+        {
+            if (!QueueAdapters.ContainsKey(typeof(T)))
+            {
+                SetupMessageType<T>(10);
+            }
+
+            await ((MessageQueueAdapter<T>)QueueAdapters[typeof(T)]).Send(message).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public void Receive<T>(Action<T> callback, int batchSize = 10)
+            where T : class
+        {
+            lock (Gate)
+            {
+                if (MessageSubs.ContainsKey(typeof(T)))
+                {
+                    throw new InvalidOperationException("You already added a callback to this message type. Only one callback per type is supported.");
+                }
+
+                SetupMessageType<T>(batchSize).StartReading();
+
+                MessageSubs.Add(
+                    typeof(T),
+                    MessagesIn.OfType<T>().Subscribe(callback));
+            }
+        }
+
+        /// <inheritdoc />
+        public void CancelReceive<T>()
+            where T : class
+        {
+            lock (Gate)
+            {
+                if (!QueueAdapters.ContainsKey(typeof(T)))
```

Comment:
You've this logic repeated a bunch of times. How about adding the below:
`AssertQueueAdapterExists<T>() => QueueAdapters.TryGetValue(typeof(T),out var result) ? result : throw ...`
which now chains nicely to the other methods as 
```
Lock<T>(T message) => AssertQueueAdapterExists<T>().Lock(message);
Complete<T>(T message) => AssertQueueAdapterExists<T>().Complete(message);
// etc
```
most of your below methods become one liners now.